radix-engine/src/model/package.rs

```diff
@@ -1,18 +1,130 @@
+use crate::engine::EnvModuleResolver;
+use crate::errors::WasmValidationError;
 use sbor::*;
+use scrypto::buffer::scrypto_decode;
+use scrypto::prelude::HashMap;
+use scrypto::rust::string::String;
+use scrypto::rust::vec;
 use scrypto::rust::vec::Vec;
+use wasmi::{
+    ExternVal, ImportsBuilder, MemoryRef, Module, ModuleInstance, ModuleRef, NopExternals,
+    RuntimeValue,
+};
 
 /// A collection of blueprints, compiled and published as a single unit.
 #[derive(Debug, Clone, TypeId, Encode, Decode)]
 pub struct Package {
     code: Vec<u8>,
+    blueprints: HashMap<String, Type>,
+}
+
+pub enum PackageError {
+    BlueprintNotFound,
 }
 
 impl Package {
-    pub fn new(code: Vec<u8>) -> Self {
-        Self { code }
+    /// Validates and creates a package
+    pub fn new(code: Vec<u8>) -> Result<Self, WasmValidationError> {
+        // Parse
+        let parsed = Self::parse_module(&code)?;
+
+        // check floating point
+        parsed
+            .deny_floating_point()
+            .map_err(|_| WasmValidationError::FloatingPointNotAllowed)?;
+
+        // Instantiate
+        let instance = ModuleInstance::new(
+            &parsed,
+            &ImportsBuilder::new().with_resolver("env", &EnvModuleResolver),
+        )
+        .map_err(|_| WasmValidationError::InvalidModule)?;
+
+        // Check start function
+        if instance.has_start() {
+            return Err(WasmValidationError::StartFunctionNotAllowed);
+        }
+        let module = instance.assert_no_start();
+
+        // Check memory export
+        let memory = match module.export_by_name("memory") {
+            Some(ExternVal::Memory(mem)) => mem,
+            _ => return Err(WasmValidationError::NoValidMemoryExport),
+        };
+
+        let rtn = module
+            .invoke_export("package_init", &[], &mut NopExternals)
+            .map_err(|e| WasmValidationError::NoPackageInitExport(e.into()))?
+            .ok_or(WasmValidationError::InvalidPackageInit)?;
+
+        let blueprints = match rtn {
+            RuntimeValue::I32(ptr) => {
+                let len: u32 = memory
+                    .get_value(ptr as u32)
+                    .map_err(|_| WasmValidationError::InvalidPackageInit)?;
+
+                // SECURITY: meter before allocating memory
+                let mut data = vec![0u8; len as usize];
+                memory
+                    .get_into((ptr + 4) as u32, &mut data)
+                    .map_err(|_| WasmValidationError::InvalidPackageInit)?;
```

Comment:
Minor: another place for direct memory read?

Reply:
yup definitely